### PR TITLE
 Improvements to the archive tab

### PR DESCRIPTION
**Pull request by antlarr**
* Fix for unmounting archives by normalizing the paths

Unmounting an archive didn't work for me because a directory like '/foo/' is compared to the items in the list of mounted directories, which contained '/foo' . This PR fixes this problem by normalizing the paths before comparison.

* Use the wants_folder parameter when asking the user for a mount point or an extraction point

* Improvements to the mount/unmount feature

Make it possible to mount more than one archive at the same time. When the user selects an archive, the mount button is updated and changes between Mount/Unmount actions depending on the state of the selected archive.

* Use a mount icon for mounted archives and check premounted archives  …
Add a column to show an icon in mounted archives.

* Use psutil to check on startup (and refreshes) the mounted archives,
so Vorta doesn't get out-of-sync with what's really mounted on the system.

* Add a context menu to the archive table (and open folder action)

Add a context menu to the archive table and a way to quickly open a file browser on the archive's mount point.

* Also open the mount point folder when double clicking on the mounted icon

### CODE REVIEW

**Comment by antlarr** [comment]
Also, I'd like to ask about some plans I have:

* Add some kind of mounted icon to the archive table when an archive is mounted, so the user can quickly see which are mounted.
* Since the dictionary of mount points per archive is stored in-process, it's lost when the user restarts Vorta. Which means that upon restart, all archives appear as unmounted and the user has to resort to the command line to unmount them, which is not nice. I would like to use psutil to parse the running commands on the system, search for "borg mount" commands and extract the archive names/mount points from there so the information used by Vorta is always correct (I would love to see any other way to get the archive associated with a mount point, but I'm afraid I couldn't find any other way to get that information but to parse command lines).

Is there interest in those changes?

**Comment by m3nu** [comment]
> Is there interest in those changes?

Yes. Definitely interested. This was neglected so far. But there is some [existing code to check for mount points](https://github.com/borgbase/vorta/blob/master/src/vorta/borg/umount.py). It's currently used to validate the mount point before unmounting it.

I've opened an issue for this feature at #131.

**Comment by m3nu** [comment]
How about merging PRs #128 and #129, since they are related? Just pulling the code. Excited to test it.

**Comment by m3nu** [comment]
Sounds good.

File: src/vorta/utils.py [hunk 1]
```diff
@@ -246,3 +246,13 @@ def format_archive_name(profile, archive_name_tpl):
...
     }
     return archive_name_tpl.format(**available_vars)
+
+
+def open_folder(folder):
+    """Open a file browser on a folder."""
+    if sys.platform == 'darwin':
+        cmd = 'open'
+    else:
+        cmd = 'xdg-open'
```
Comment (ThomasWaldmann) [inline src/vorta/utils.py]: cmd = 'open' if sys.platform == 'darwin' else 'xdg-open'

File: src/vorta/views/archive_tab.py [hunk 2]
```diff
@@ -181,64 +194,74 @@ def mount_action(self):
...
+            self.update_mount_button_text()
 
     def umount_action(self):
-        if self.mount_point is not None:
+        snapshot_name = self.selected_archive_name()
+
+        try:
+            mount_point = self.mount_points[snapshot_name]
+        except KeyError:
+            mount_point = None
```
Comment (ThomasWaldmann) [inline src/vorta/views/archive_tab.py]: mount_point = self.mount_points.get(snapshot_name)

File: src/vorta/views/archive_tab.py [hunk 3]
```diff
@@ -297,8 +320,71 @@ def receive():
...
+
+    def update_mount_button_text(self):
+        snapshot_name = self.selected_archive_name()
+        if not snapshot_name:
+            return
+
+        if snapshot_name in self.mount_points:
+            self.set_mount_button_mode('Unmount')
+        else:
+            self.set_mount_button_mode('Mount')
```
Comment (ThomasWaldmann) [inline src/vorta/views/archive_tab.py]: mode = 'Unmount' if snapshot_name in self.mount_points else 'Mount'
self.set_mount_button_mode(mode)

File: src/vorta/views/archive_tab.py [hunk 4]
```diff
@@ -297,8 +320,71 @@ def receive():
...
+
+    def open_folder_action(self):
+        snapshot_name = self.selected_archive_name()
+        if not snapshot_name:
+            return
+
+        try:
+            mount_point = self.mount_points[snapshot_name]
+        except KeyError:
+            mount_point = None
```
Comment (ThomasWaldmann) [inline src/vorta/views/archive_tab.py]: .get(), see above.

File: src/vorta/views/archive_tab.py [hunk 5]
```diff
@@ -297,8 +320,71 @@ def receive():
...
+
+        if mount_point:
+            open_folder(mount_point)
```
Comment (ThomasWaldmann) [inline src/vorta/views/archive_tab.py]: above you used "is not None", maybe always use the same method.

if the empty string or any other falsy value except None can not happen, the shorter one is nicer.

File: src/vorta/utils.py (hunk 1 above)
Comment (antlarr) [reply src/vorta/utils.py]: done

File: src/vorta/views/archive_tab.py (hunk 2 above)
Comment (antlarr) [reply src/vorta/views/archive_tab.py]: sure

File: src/vorta/views/archive_tab.py (hunk 3 above)
Comment (antlarr) [reply src/vorta/views/archive_tab.py]: ok

File: src/vorta/views/archive_tab.py (hunk 4 above)
Comment (antlarr) [reply src/vorta/views/archive_tab.py]: ok

File: src/vorta/views/archive_tab.py (hunk 5 above)
Comment (antlarr) [reply src/vorta/views/archive_tab.py]: I tend to use the implicit check in my code but I noticed Vorta sources tend to prefer the "is not None" kind of check, so I tried to use it everywhere to keep the code style but this one seem to have slipped. I'll fix it. Thanks for the review!

File: src/vorta/views/archive_tab.py [hunk 6]
```diff
@@ -181,11 +196,20 @@ def list_result(self, result):
     def selected_archive_name(self):
         row_selected = self.archiveTable.selectionModel().selectedRows()
         if row_selected:
-            snapshot_cell = self.archiveTable.item(row_selected[0].row(), 3)
+            snapshot_cell = self.archiveTable.item(row_selected[0].row(), 4)
```
Comment (ThomasWaldmann) [inline src/vorta/views/archive_tab.py]: iirc @m3nu just removed the word "snapshot" from the code in the i18n PR and replaced it with "archive".

Comment (m3nu) [reply src/vorta/views/archive_tab.py]: There will be some conflicts with #134. If we merge #134 before #129, it should be rebased when ready.

**Comment by m3nu** [comment]
Resolved some conflicts with #134. Are you still planning changes, @antlarr ?

**Comment by m3nu** [comment]
One test and flake8 failing.

File: src/vorta/utils.py [hunk 7]
```diff
@@ -252,3 +253,22 @@ def open_folder(folder):
     """Open a file browser on a folder."""
     cmd = 'open' if sys.platform == 'darwin' else 'xdg-open'
     subprocess.Popen([cmd, folder])
+
+
+def get_mount_points(repo_url):
+    processes = [p for p in psutil.process_iter() if 'borg' in p.name() and 'mount' in p.cmdline()]
+
+    mount_points = {}
+    for proc in psutil.process_iter():
```
[the remark this answers is not in the document]
Comment (antlarr) [reply src/vorta/utils.py]: Sorry, that was code from an early stage of development. I just removed it

**Comment by antlarr** [comment]
I'll fix those comments by Thomas and the test, then I'm done. I'm just not sure when I'll have time (probably tomorrow, but I'll try to finish everything today). 

Btw, what flake8 issue do you mean?

File: src/vorta/views/archive_tab.py (hunk 6 above)
Comment (antlarr) [reply src/vorta/views/archive_tab.py]: I renamed the existing snapshot_* variables to archive_*

File: src/vorta/views/archive_tab.py [hunk 8]
```diff
@@ -366,17 +389,32 @@ def archiveTable_context_menu_event(self, event):
...
+        if column == 3:
+            self.open_folder_action()
+
+    def row_of_archive(self, archive_name):
+        items = self.archiveTable.findItems(archive_name, QtCore.Qt.MatchExactly)
+        rows = [item.row() for item in items if item.column() == 4]
+        if rows == []:
+            return None
+
+        return rows[0]
```
[the remark this answers is not in the document]
Comment (antlarr) [reply src/vorta/views/archive_tab.py]: ok

File: src/vorta/views/archive_tab.py [hunk 9]
```diff
@@ -181,11 +196,20 @@ def list_result(self, result):
...
 
+    def set_mount_button_mode(self, mode):
+        self.mountButton.clicked.disconnect()
+        if mode == 'Mount':
+            self.mountButton.setText('Mount')
+            self.mountButton.clicked.connect(self.mount_action)
+        elif mode == 'Unmount':
+            self.mountButton.setText('Unmount')
+            self.mountButton.clicked.connect(self.umount_action)
+
```
[the remark this answers is not in the document]
Comment (antlarr) [reply src/vorta/views/archive_tab.py]: done

File: src/vorta/utils.py [hunk 10]
```diff
@@ -252,3 +253,22 @@ def open_folder(folder):
...
+        if proc.name() != 'borg':
+            continue
+
+        if 'mount' not in proc.cmdline():
+            continue
+
+        if proc.cmdline()[-2].startswith(repo_url):
+            archive_name = proc.cmdline()[-2][len(repo_url) + 2:]
+            mount_point = proc.cmdline()[-1]
+            mount_points[archive_name] = mount_point
```
[the remark this answers is not in the document]
Comment (antlarr) [reply src/vorta/utils.py]: I changed a bit the algorithm so the archive_name position in the command line is not hardcoded anymore

**Comment by m3nu** [comment]
Just tested this. Good stuff really. Made 2 small changes:

- detecting mount points via the `borg` binary only works when using the fat binary. The process name could also be `pythonxxx`.
- I put the truncated mount point in the table. Why not show it, if it's known?

Now I just need to figure out how to merge this, since there are some extra commits from yesterday in here.

**Comment by m3nu** [comment]
Moved a squashed version to #156 . Easier to see the changes there.

**Pull request closed without merging**